File: aecp-python/src/aecp/quality/metrics.py
"""Quality metrics for embedding-space mappings.

Cosine similarity alone is misleading for migration decisions. Prefer
top-k retrieval retention: the fraction of queries whose true nearest
neighbor is preserved after mapping.
"""

from __future__ import annotations

from typing import Any

import numpy as np

from aecp.mapping.base import l2_normalize


def _spearman_rho(a: np.ndarray, b: np.ndarray) -> float:
    """Spearman rank correlation without a scipy dependency."""
    if a.size < 2:
        return 0.0
    ra = np.argsort(np.argsort(a)).astype(np.float64)
    rb = np.argsort(np.argsort(b)).astype(np.float64)
    ra -= ra.mean()
    rb -= rb.mean()
    denom = float(np.sqrt(np.sum(ra**2) * np.sum(rb**2)))
    if denom < 1e-12:
        return 0.0
    return float(np.sum(ra * rb) / denom)


def cosine_similarity(a: np.ndarray, b: np.ndarray) -> float:
    """Cosine similarity between two vectors."""
    a = np.asarray(a, dtype=np.float64).ravel()
    b = np.asarray(b, dtype=np.float64).ravel()
    if a.size != b.size:
        raise ValueError(f"Size mismatch: {a.size} vs {b.size}")
    na = float(np.linalg.norm(a))
    nb = float(np.linalg.norm(b))
    if na < 1e-12 or nb < 1e-12:
        return 0.0
    return float(np.clip(np.dot(a, b) / (na * nb), -1.0, 1.0))


def pairwise_cosine_stats(
    predicted: np.ndarray,
    target: np.ndarray,
) -> dict[str, float]:
    """Row-wise cosine between predicted and target; return mean/median/p5."""
    predicted = np.asarray(predicted, dtype=np.float64)
    target = np.asarray(target, dtype=np.float64)
    if predicted.shape != target.shape:
        raise ValueError(
            f"Shape mismatch: {predicted.shape} vs {target.shape}"
        )
    p = l2_normalize(predicted)
    t = l2_normalize(target)
    sims = np.sum(p * t, axis=1)
    return {
        "mean": float(np.mean(sims)),
        "median": float(np.median(sims)),
        "p5": float(np.percentile(sims, 5)),
        "min": float(np.min(sims)),
        "max": float(np.max(sims)),
    }


def topk_retention(
    queries: np.ndarray,
    corpus: np.ndarray,
    k: int = 1,
) -> float:
    """Fraction of rows whose true match (same index) is in the top-k neighbors.

    For each query ``queries[i]``, retrieve top-k nearest neighbors in
    ``corpus`` by cosine similarity and check whether index ``i`` appears.
    Self-match exclusion is not applied: the intended neighbor *is* index i.
    """
    queries = l2_normalize(np.asarray(queries, dtype=np.float64))
    corpus = l2_normalize(np.asarray(corpus, dtype=np.float64))
    n = queries.shape[0]
    if n == 0:
        return 0.0
    k = min(k, n)
    sims = queries @ corpus.T
    # argsort ascending; take last k
    top = np.argpartition(sims, -k, axis=1)[:, -k:]
    correct = 0
    for i in range(n):
        if i in top[i]:
            correct += 1
    return float(correct) / float(n)


def rank_correlation(
    queries_a: np.ndarray,
    corpus_a: np.ndarray,
    queries_b: np.ndarray,
    corpus_b: np.ndarray,
    sample_queries: int | None = 50,
    seed: int = 0,
) -> dict[str, float]:
    """Spearman rank correlation of neighbor rankings between two spaces."""
    rng = np.random.default_rng(seed)
    n = queries_a.shape[0]
    idx = np.arange(n)
    if sample_queries is not None and sample_queries < n:
        idx = rng.choice(n, size=sample_queries, replace=False)

    qa = l2_normalize(queries_a[idx])
    ca = l2_normalize(corpus_a)
    qb = l2_normalize(queries_b[idx])
    cb = l2_normalize(corpus_b)

    sims_a = qa @ ca.T
    sims_b = qb @ cb.T
    rhos: list[float] = []
    for i in range(len(idx)):
        rho = _spearman_rho(sims_a[i], sims_b[i])
        if np.isfinite(rho):
            rhos.append(float(rho))
    if not rhos:
        return {"mean_spearman": 0.0, "n": 0.0}
    return {"mean_spearman": float(np.mean(rhos)), "n": float(len(rhos))}


def holdout_rank_correlation(
    mapped: np.ndarray,
    target: np.ndarray,
    sample_size: int | None = 200,
    seed: int = 0,
) -> float:
    """Spearman rank correlation of pairwise similarity matrices.

    Compares the neighbor structure of mapped vectors vs true target vectors.
    Captures neighborhood-structure preservation better than cosine alone.
    """
    rng = np.random.default_rng(seed)
    n = mapped.shape[0]
    if n < 3:
        return 0.0
    if sample_size is not None and sample_size < n:
        idx = rng.choice(n, size=sample_size, replace=False)
    else:
        idx = np.arange(n)

    m = l2_normalize(mapped[idx])
    t = l2_normalize(target[idx])

    # Pairwise cosine similarities (upper triangle only for efficiency)
    sim_mapped = m @ m.T
    sim_target = t @ t.T

    # Extract upper triangle (excluding diagonal)
    triu_idx = np.triu_indices(len(idx), k=1)
    a = sim_mapped[triu_idx]
    b = sim_target[triu_idx]

    return _spearman_rho(a, b)


def mrr_delta(
    queries_mapped: np.ndarray,
    corpus_target: np.ndarray,
    queries_true: np.ndarray,
) -> dict[str, float]:
    """Mean reciprocal rank for true-index retrieval: mapped vs true queries."""

    def _mrr(queries: np.ndarray, corpus: np.ndarray) -> float:
        q = l2_normalize(queries)
        c = l2_normalize(corpus)
        sims = q @ c.T
        n = sims.shape[0]
        rr = 0.0
        for i in range(n):
            order = np.argsort(-sims[i])
            rank = int(np.where(order == i)[0][0]) + 1
            rr += 1.0 / rank
        return rr / n if n else 0.0

    mrr_mapped = _mrr(queries_mapped, corpus_target)
    mrr_true = _mrr(queries_true, corpus_target)
    return {
        "mrr_mapped": float(mrr_mapped),
        "mrr_true": float(mrr_true),
        "mrr_delta": float(mrr_mapped - mrr_true),
    }


def retrieval_retention_report(
    mapped: np.ndarray,
    target: np.ndarray,
) -> dict[str, Any]:
    """Bundle of migration-relevant quality metrics."""
    cos = pairwise_cosine_stats(mapped, target)
    return {
        "cosine": cos,
        "top1_retention": topk_retention(mapped, target, k=1),
        "top10_retention": topk_retention(
            mapped, target, k=min(10, len(target))
        ),
        "mrr": mrr_delta(mapped, target, target),
    }

File: isotrieve-npm/packages/core/src/math/metrics.ts
/**
 * Quality metrics for embedding-space mappings.
 * Cosine alone is misleading — prefer top-k retention.
 */

import { cosineSimilarity, transpose, matrixMultiply, vecNorm } from './linalg';
import { l2Normalize } from './normalize';

// ── Pairwise Cosine Stats ────────────────────────────────────────

/**
 * Row-wise cosine between predicted and target; return mean/median/p5.
 */
export function pairwiseCosineStats(
  predicted: Float64Array[],
  target: Float64Array[],
): { mean: number; median: number; p5: number; min: number; max: number } {
  if (predicted.length !== target.length) {
    throw new Error(
      `Shape mismatch: ${predicted.length} vectors vs ${target.length} vectors`,
    );
  }
  const n = predicted.length;
  if (n === 0) return { mean: 0, median: 0, p5: 0, min: 0, max: 0 };

  const sims = new Float64Array(n);
  for (let i = 0; i < n; i++) {
    sims[i] = cosineSimilarity(predicted[i], target[i]);
  }

  const sorted = Float64Array.from(sims).sort();
  let sum = 0;
  for (let i = 0; i < n; i++) sum += sims[i];

  return {
    mean: sum / n,
    median: sorted[Math.floor(n / 2)],
    p5: sorted[Math.floor(n * 0.05)],
    min: sorted[0],
    max: sorted[n - 1],
  };
}

// ── Top-K Retention ──────────────────────────────────────────────

/**
 * Fraction of rows whose true match (same index) is in top-k neighbors.
 * Uses argpartition-like selection for efficiency.
 */
export function topkRetention(
  queries: Float64Array[],
  corpus: Float64Array[],
  k = 1,
): number {
  const n = queries.length;
  if (n === 0) return 0;
  k = Math.min(k, n);

  const sims = matrixMultiply(queries, transpose(corpus));

  let correct = 0;
  for (let i = 0; i < n; i++) {
    // Quickselect top-k
    const row = sims[i];
    const topK = quickSelectTopK(row, k);
    if (topK.has(i)) correct++;
  }
  return correct / n;
}

/**
 * Find the indices of the top-k largest values using partial sort.
 * Returns a Set of indices.
 */
function quickSelectTopK(arr: Float64Array, k: number): Set<number> {
  const n = arr.length;
  if (k >= n) {
    return new Set(Array.from({ length: n }, (_, i) => i));
  }

  // Use a min-heap of size k
  const heap: Array<{ val: number; idx: number }> = [];

  for (let i = 0; i < n; i++) {
    if (heap.length < k) {
      heap.push({ val: arr[i], idx: i });
      if (heap.length === k) heapifyUp(heap, 0);
    } else if (arr[i] > heap[0].val) {
      heap[0] = { val: arr[i], idx: i };
      heapifyDown(heap, 0, k);
    }
  }

  return new Set(heap.map((h) => h.idx));
}

function heapifyUp(heap: Array<{ val: number; idx: number }>, i: number): void {
  while (i > 0) {
    const parent = (i - 1) >> 1;
    if (heap[i].val < heap[parent].val) {
      [heap[i], heap[parent]] = [heap[parent], heap[i]];
      i = parent;
    } else break;
  }
}

function heapifyDown(
  heap: Array<{ val: number; idx: number }>,
  i: number,
  size: number,
): void {
  while (true) {
    let smallest = i;
    const left = 2 * i + 1;
    const right = 2 * i + 2;
    if (left < size && heap[left].val < heap[smallest].val) smallest = left;
    if (right < size && heap[right].val < heap[smallest].val) smallest = right;
    if (smallest === i) break;
    [heap[i], heap[smallest]] = [heap[smallest], heap[i]];
    i = smallest;
  }
}

// ── Spearman Rank Correlation ────────────────────────────────────

/**
 * Spearman rank correlation without external dependencies.
 */
export function spearmanRho(a: Float64Array, b: Float64Array): number {
  if (a.length < 2) return 0;
  const n = a.length;

  const rankA = rank(a);
  const rankB = rank(b);

  let sumA = 0;
  let sumB = 0;
  for (let i = 0; i < n; i++) {
    sumA += rankA[i];
    sumB += rankB[i];
  }
  const meanA = sumA / n;
  const meanB = sumB / n;

  let ssA = 0;
  let ssB = 0;
  let ab = 0;
  for (let i = 0; i < n; i++) {
    const da = rankA[i] - meanA;
    const db = rankB[i] - meanB;
    ssA += da * da;
    ssB += db * db;
    ab += da * db;
  }

  const denom = Math.sqrt(ssA * ssB);
  if (denom < 1e-12) return 0;
  return ab / denom;
}

function rank(arr: Float64Array): Float64Array {
  const n = arr.length;
  const indexed = Array.from({ length: n }, (_, i) => ({ val: arr[i], idx: i }));
  indexed.sort((a, b) => a.val - b.val);

  const result = new Float64Array(n);
  let i = 0;
  while (i < n) {
    let j = i;
    while (j < n - 1 && indexed[j + 1].val === indexed[j].val) j++;
    const avgRank = (i + j) / 2 + 1; // 1-indexed
    for (let k = i; k <= j; k++) {
      result[indexed[k].idx] = avgRank;
    }
    i = j + 1;
  }
  return result;
}

// ── Holdout Rank Correlation ─────────────────────────────────────

/**
 * Spearman rank correlation of pairwise similarity matrices.
 * Compares neighbor structure of mapped vs true target vectors.
 */
export function holdoutRankCorrelation(
  mapped: Float64Array[],
  target: Float64Array[],
  sampleSize: number | null = 200,
  seed = 0,
): number {
  const n = mapped.length;
  if (n < 3) return 0;

  // Sample indices
  let idx: number[];
  if (sampleSize !== null && sampleSize < n) {
    idx = sampleIndices(n, sampleSize, seed);
  } else {
    idx = Array.from({ length: n }, (_, i) => i);
  }

  const m = l2Normalize(idx.map((i) => mapped[i])) as Float64Array[];
  const t = l2Normalize(idx.map((i) => target[i])) as Float64Array[];

  const simMapped = matrixMultiply(m, transpose(m));
  const simTarget = matrixMultiply(t, transpose(t));

  const len = idx.length;
  const aVals: number[] = [];
  const bVals: number[] = [];

  // Upper triangle only
  for (let i = 0; i < len; i++) {
    for (let j = i + 1; j < len; j++) {
      aVals.push(simMapped[i][j]);
      bVals.push(simTarget[i][j]);
    }
  }

  return spearmanRho(new Float64Array(aVals), new Float64Array(bVals));
}

function sampleIndices(n: number, k: number, seed: number): number[] {
  // Deterministic sampling
  let s = seed | 0;
  const sampled = new Set<number>();
  while (sampled.size < k && sampled.size < n) {
    s = (s * 1103515245 + 12345) & 0x7fffffff;
    sampled.add(s % n);
  }
  return Array.from(sampled);
}

// ── MRR Delta ────────────────────────────────────────────────────

/**
 * Mean reciprocal rank for true-index retrieval: mapped vs true queries.
 */
export function mrrDelta(
  mappedQueries: Float64Array[],
  corpus: Float64Array[],
  trueQueries: Float64Array[],
): { mrrMapped: number; mrrTrue: number; mrrDelta: number } {
  const mrrMapped = computeMrr(mappedQueries, corpus);
  const mrrTrue = computeMrr(trueQueries, corpus);
  return {
    mrrMapped,
    mrrTrue,
    mrrDelta: mrrMapped - mrrTrue,
  };
}

function computeMrr(queries: Float64Array[], corpus: Float64Array[]): number {
  const n = queries.length;
  if (n === 0) return 0;

  const sims = matrixMultiply(queries, transpose(corpus));
  let rr = 0;
  for (let i = 0; i < n; i++) {
    const row = sims[i];
    let bestRank = 0;
    let bestSim = -Infinity;
    for (let j = 0; j < row.length; j++) {
      if (row[j] > bestSim) {
        bestSim = row[j];
        bestRank = j;
      }
    }
    // Find actual rank of index i
    let rank = 1;
    for (let j = 0; j < row.length; j++) {
      if (row[j] > row[i]) rank++;
    }
    rr += 1.0 / rank;
  }
  return rr / n;
}

// ── Bundle Report ────────────────────────────────────────────────

/**
 * Bundle of migration-relevant quality metrics.
 */
export function retrievalRetentionReport(
  mapped: Float64Array[],
  target: Float64Array[],
): {
  cosine: { mean: number; median: number; p5: number; min: number; max: number };
  top1Retention: number;
  top10Retention: number;
  mrr: { mrrMapped: number; mrrTrue: number; mrrDelta: number };
} {
  const normalizedMapped = l2Normalize(mapped) as Float64Array[];
  const normalizedTarget = l2Normalize(target) as Float64Array[];
  return {
    cosine: pairwiseCosineStats(normalizedMapped, normalizedTarget),
    top1Retention: topkRetention(normalizedMapped, normalizedTarget, 1),
    top10Retention: topkRetention(normalizedMapped, normalizedTarget, Math.min(10, target.length)),
    mrr: mrrDelta(normalizedMapped, normalizedTarget, normalizedTarget),
  };
}
